### ISSUE TITLE
Screenshare regression on Sway+Pipewire+xdg-desktop-portal-wlr ?

### ISSUE THREAD

**Issue by stacyharper** [qutebrowser/qutebrowser#6752]
**Version info**:

qutebrowser v2.3.1
Git commit: 
Backend: QtWebEngine 5.15.2, based on Chromium 87.0.4280.144
Qt: 5.15.2

**Does the bug happen if you start with `--temp-basedir`?**: Yes it does !

**Description**

I'm trying web rtc screen sharing with : https://mozilla.github.io/webrtc-landing/gum_test.html

It got the working Sway+Pipewire+xdg-desktop-portal-wlr setup.

I see this :( https://tube.missbanal.net/videos/watch/d7fdcc0f-3bb4-4df3-a453-a02c1299cd03

What we dont really see is the slurp command to select the current workspace.

So in short, everything start as if it works, but it does not. When using google meets by example, people see a black screen.

This exact same setup works on firefox.

**Comment by The-Compiler** [comment]
I mean if you try with KDE Falkon and that environment variable set.

**Comment by The-Compiler** [comment]
Ah, right, that's unfortunate... I suppose I need to start to ask people to test things with simplebrowser instead indeed...

**Comment by maximbaz** [comment]
I'm a bit suspicious about the hypothesis of Chromium version because this _used to_ work in qutebrowser, for quite a long time actually, and I don't think there was a set of Chromium versions where screensharing stopped working... 

It doesn't work in `qt6-test` branch either by the way, but it looks different, in `qt6-branch` it looks as if I am screensharing an empty desktop with cursor in the middle, while in latest released qutebrowser it looks like on the screenshots above, as if it doesn't start at all.

My bet would rather be on pipewire breaking something recently, xdg-desktop-portal-wlr didn't have a new release for a long while (until today, but with new 0.5.0 version still not working)...

**Comment by stacyharper** [comment]
> My bet would rather be on pipewire breaking something recently, xdg-desktop-portal-wlr didn't have a new release for a long while (until today, but with new 0.5.0 version still not working)...

A pipewire regression while it still works on firefox, obs, etc ?

**Comment by tardypad** [comment]
Here's my experience with this issue (using Arch Linux) AFAIR:

- it used to work fine on qutebrowser and Chromium
- I have a pipewire update (I'm quite confident this was to version 0.3.33) then screen sharing breaks on both on qutebrowser and chromium. So I downgrade and block the updates.
- l'm retesting with every qutebrowser, pipewire, libpipewire02, xdg-desktop-portal-wlr,... updates that come
- with pipewire 0.3.35 or 0.3.36 (not quite sure), the screen sharing works again on chromium but not on qutebrowser
- this is still my current experience today with pipewire 0.3.39 (pipewire-media-session 0.4.0 for some reasons)

Here is the error I see in the journal when trying to screenshare:
```
pipewire[4668]: mod.protocol-native: old version detected
pipewire[4668]: mod.protocol-native: 0x558ac2157300: client 0x558ac225f190 error -95 (Operation not supported)
```

I regret now not reporting this bug immediately as it came but I wasn't sure I didn't modify anything else on my config.

**Comment by maximbaz** [comment]
You have a good point @stacyharper. I've also tried to run qutebrowser from the older tag, unless I'm doing something wrong the screensharing is not working either...

What do you guys think of trying to ask folks over at  `xdg-desktop-portal-wlr` repo? In my experience those guys are most helpful, and with initial set of good debug info I hope they'd help to narrow the issue down. I'm unfortunately a bit short on time now, does anyone have opportunity to make a detailed bug report over there and ask for help?

**Comment by heittpr** [comment]
Any updates on this? I did some more testing today and found out that qutebrowser now segfaults on my machine when trying to screenshare using xdg-desktop-portal-wlr.

I suspect this has nothing to do with version `v2.4.0` but actually with some Qt/PyQt update (or mabye something else entirely) since a fresh `v2.3.1` copy [installed in a virtualenv](https://github.com/qutebrowser/qutebrowser/blob/master/doc/install.asciidoc#installing-qutebrowser-with-virtualenv) with `--pyqt-type link` still segfaults.

Here are the relevants stacktraces:

```
#0  0x00007fff84407d62 in ?? () from /usr/lib64/pipewire-0.2/libpipewire-module-protocol-native.so
#1  0x00007fff8440d4d2 in ?? () from /usr/lib64/pipewire-0.2/libpipewire-module-protocol-native.so
#2  0x00007fff8441bbee in ?? () from /usr/lib64/spa/support/libspa-support.so
#3  0x00007fff6873c65d in ?? () from /usr/lib/libpipewire-0.2.so.1
#4  0x00007ffff7a7beae in start_thread (arg=0x7ffefcff9640) at pthread_create.c:463
#5  0x00007ffff7b922ff in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by The-Compiler** [comment]
Well that shows that the crash is inside libpipewire at least. Not much qutebrowser can do I guess.

**Comment by tinywrkb** [comment]
I also hit this issue recently after updating my system.  
It's likely `PipeWire 0.3.x` or `xdg-desktop-portal` regression.  
It's crashing only with libpipewire 0.2. If I switch to WebEngine 6.2.2 with its `libpipewire 0.3` support, then it's working perfectly.

Somehow it's also broken, though not crashing, with recent Chromium and Chrome. Could be a different issue considering `libpipewire 0.3` is used. Maybe be related to the recent DMABUF support.  
It's not worth looking at before Sway 1.7 is released.

No problem with OBS, so the `libpipewire 0.3` issue is limited to recent Chromium versions.

**Comment by amydotsh** [comment]
these are the only 2 errors i get on gum test, regardless of `--qt-flag enable-features=WebRTCPipeWireCapturer`
<details>
<summary>errors</summary>

```
~%12:19:31 WARNING: Unable to link XRandR monitor functions.                                
[5164:5164:0201/121932.975961:ERROR:base_capturer_pipewire.cc(279)] Failed to load the PipeWire library and symbols.
```
</details>

xdg-desktop-portal and xdg-desktop-portal-wlr are running, in riverwm's init file i have dbus-update-activation-environment WAYLAND_DISPLAY XDG_CURRENT_DESKTOP=river
the gum test works on firefox and chromium, so maybe it has to do with [this note](https://wiki.archlinux.org/title/PipeWire#WebRTC_screen_sharing)?
<details>
<summary>note</summary>
Note: xdg-desktop-portal 1.10.0 fixed a mismatch between specification and implementation of its D-Bus interface. [2] Hence, some clients may not work with xdg-desktop-portal 1.10.0 or newer. Current progress for OBS Studio, Chromium and Firefox.<br>
Firefox (84+) supports this method by default, while on Chromium (73+) one needs to enable WebRTC PipeWire support by setting the corresponding (experimental) flag at the URL `chrome://flags/#enable-webrtc-pipewire-capturer`.
</details>

**Comment by antonio-rojas** [comment]
Arch users can try qt5-webengine 5.15.8-9. It is patched for pipewire 0.3 and seems to work fine for me.

**Comment by tinywrkb** [comment]
Thanks @antonio-rojas for porting the PipeWire 0.3 changes, that's great work!  
I will try your patch with the QtWebEngine Flatpak base app, and send a PR later to have this working also with QtWebEngine Flatpak apps.

**Comment by tinywrkb** [comment]
I can confirm that the patch is working also with the qutebrowser Flatpak app. It will be fixed on Flathub a bit after the QtWebEngine base app changes will be merged.

**Comment by tinywrkb** [comment]
The Flathub app is now updated.

**Comment by nrdxp** [comment]
Just leaving a comment here that I cannot get screensharing to work at all with qutebrowser using the qt6 build along with the [vivarium](https://github.com/inclement/vivarium) wayland compositor.

I have already confirmed that webengine is using pipewire 3, and I can use screensharing with chromium if I pass the `--enable-features=WebRTCPipeWireCapturer` flag. Not sure what else I can try :shrug:

**Comment by foxyseta** [comment]
Hi! Same problem here. The "getUserMedia / getDisplayMedia Test Page" gives me a "Success" message, but a stale cursor on a black screen is shown instead. Chromium works fine instead, but only on window and tab (not screen) sharing. My setup is Wayland + Sway + QuteBrowser, too.

```bash
$ qutebrowser --temp-basedir --qt-flag enable-features=WebRTCPipeWireCapturer
16:19:09 WARNING: Unable to link XRandR monitor functions.
[19295:19329:0202/161909.954579:ERROR:screen_capturer_x11.cc(176)] Unable to link XRandR monitor functions.
```

```bash
$ echo $XDG_CURRENT_DESKTOP 
sway
$ echo $XDG_SESSION_TYPE
wayland
```

```
$ qutebrowser --version
         ______     ,,
    ,.-"`      | ,-` |
  .^           ||    |
 /    ,-*^|    ||    |
;    /    |    ||    ;-*```^*.
;   ;     |    |;,-*`         \
|   |     |  ,-*`    ,-"""\    \
|    \   ,-"`    ,-^`|     \    |
 \    `^^    ,-;|    |     ;    |
  *;     ,-*`  ||    |     /   ;;
    `^^`` |    ||    |   ,^    /
          |    ||    `^^`    ,^
          |  _,"|        _,-"
          -*`   ****"""``

qutebrowser v2.5.0
Git commit: 
Backend: QtWebEngine 5.15.3, based on Chromium 87.0.4280.144
Qt: 5.15.4 (compiled 5.15.3)

CPython: 3.10.9
PyQt: 5.15.6

sip: 6.1.0.dev2104271705
colorama: no
jinja2: 3.0.3
pygments: 2.11.2
yaml: 6.0
adblock: 0.5.2
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.5
PyQt5.QtWebKitWidgets: no
pdf.js: 2.6.347 (/usr/share/pdf.js/build/pdf.js)
sqlite: 3.38.5
QtNetwork SSL: OpenSSL 1.1.1s  1 Nov 2022

Style: QFusionStyle
Platform plugin: wayland
OpenGL: Intel, 4.6 (Compatibility Profile) Mesa 21.3.8 (git-53b2b224dc)
Platform: Linux-5.15.90-0-lts-x86_64-with, 64bit
Linux distribution: Alpine Linux v3.16 (alpine)
Frozen: False
Imported from /usr/lib/python3.10/site-packages/qutebrowser
Using Python from /usr/bin/python3
Qt library executable path: /usr/lib/qt5/libexec, data path: /usr/share/qt5

Paths:
cache: /home/foxy/.cache/qutebrowser
config: /home/foxy/.config/qutebrowser
data: /home/foxy/.local/share/qutebrowser
runtime: /tmp/1000-runtime-dir/qutebrowser
system data: /usr/share/qutebrowser

Autoconfig loaded: yes
Config.py: no config.py was loaded
Uptime: 0:00:00
```

```
$ pipewire --version
pipewire
Compiled with libpipewire 0.3.51
Linked with libpipewire 0.3.51
```

**Comment by stacyharper** [comment]
I don't think I still have this very old issue. I'm closing this
